File: src/lib/watermark/fujifilm-classic.ts
import { normalizeModel } from "@/lib/exif/brand";
import type { FrameParams, TemplateConfig } from "@/stores/types";
import {
  cleanDisplayText,
  formatTakenAt,
  getPreviewFontFamily,
  sanitizeCustomLines,
} from "./classic-bottom";
import {
  dprFor,
  fitPhoto,
  getCanvasRatio,
  paramsLine,
  type ShowcasePhoto,
} from "./renderer-utils";

const FUJI_GREEN = "#00643f";
const FUJI_PAPER = "#f5f1e7";
const FUJI_INK = "#161616";
const FUJI_MUTED = "#5a5650";

/**
 * Fujifilm classic style. Cream paper background; photo sits on top with a
 * left "FUJIFILM" green wordmark + film simulation pill on the bottom-left
 * and EXIF / date column on the bottom-right.
 */
export function drawFujifilmClassicPreview(
  canvas: HTMLCanvasElement,
  image: HTMLImageElement,
  _logo: HTMLImageElement | null,
  photo: ShowcasePhoto,
  frameParams: FrameParams,
  config: TemplateConfig,
) {
  const ctx = canvas.getContext("2d");
  if (!ctx) return;

  const dpr = dprFor(canvas);
  const baseWidth = 900;
  const ratio = getCanvasRatio(frameParams.canvasRatio, frameParams.canvasOrientation);
  const canvasW = baseWidth;
  const canvasH = canvasW / ratio;
  const watermarkActive = config.showWatermark ?? true;

  const sideMargin = canvasW * 0.05;
  const topMargin = canvasH * 0.05;
  const captionH = watermarkActive ? Math.max(96, canvasH * 0.16) : canvasH * 0.05;

  const photoArea = {
    x: sideMargin,
    y: topMargin,
    w: canvasW - sideMargin * 2,
    h: canvasH - topMargin - captionH,
  };
  const placed = fitPhoto(photoArea, photo.width, photo.height);

  canvas.width = Math.round(canvasW * dpr);
  canvas.height = Math.round(canvasH * dpr);
  canvas.style.aspectRatio = `${canvasW} / ${canvasH}`;
  ctx.setTransform(dpr, 0, 0, dpr, 0, 0);
  ctx.clearRect(0, 0, canvasW, canvasH);

  ctx.fillStyle = FUJI_PAPER;
  ctx.fillRect(0, 0, canvasW, canvasH);
  ctx.drawImage(image, placed.x, placed.y, placed.w, placed.h);

  if (!watermarkActive) return;

  const fontFamily = getPreviewFontFamily(frameParams.fontFamily);
  const captionTop = placed.y + placed.h + canvasH * 0.03;

  // Left: FUJIFILM wordmark + film simulation pill.
  ctx.textBaseline = "alphabetic";
  ctx.textAlign = "left";
  const wordmarkSize = Math.max(20, frameParams.fontSize * 1.7);
  ctx.fillStyle = FUJI_GREEN;
  ctx.font = `900 ${wordmarkSize}px ${fontFamily}`;
  ctx.fillText("FUJIFILM", placed.x, captionTop + wordmarkSize * 0.85);

  const wordWidth = ctx.measureText("FUJIFILM").width;

  // Film simulation pill (CLASSIC CHROME by default — falls back to first custom line).
  const customLines = sanitizeCustomLines(config.customLines);
  const sim = (customLines[0] || "CLASSIC CHROME").toUpperCase();
  const pillSize = Math.max(10, frameParams.fontSize - 1);
  ctx.font = `700 ${pillSize}px ${fontFamily}`;
  const pillW = ctx.measureText(sim).width + pillSize * 1.4;
  const pillH = pillSize * 1.9;
  const pillX = placed.x + wordWidth + pillSize * 0.9;
  const pillY = captionTop + wordmarkSize * 0.85 - pillH * 0.78;
  ctx.fillStyle = FUJI_GREEN;
  drawPill(ctx, pillX, pillY, pillW, pillH);
  ctx.fillStyle = "#ffffff";
  ctx.textBaseline = "middle";
  ctx.textAlign = "center";
  ctx.fillText(sim, pillX + pillW / 2, pillY + pillH / 2 + 1);

  // Lens / camera below wordmark.
  const camera = config.showCamera
    ? normalizeModel(photo.exif.camera.make, photo.exif.camera.model)
    : "";
  const lens = config.showLens ? cleanDisplayText(photo.exif.lens) : "";
  const detail = [camera, lens].filter(Boolean).join("  ·  ");
  if (detail) {
    ctx.textAlign = "left";
    ctx.textBaseline = "alphabetic";
    ctx.fillStyle = FUJI_MUTED;
    ctx.font = `500 ${Math.max(11, frameParams.fontSize)}px ${fontFamily}`;
    ctx.fillText(detail, placed.x, captionTop + wordmarkSize * 0.85 + Math.max(20, frameParams.fontSize * 1.6));
  }

  // Right: params + date stacked, right-aligned with photo.
  const rightX = placed.x + placed.w;
  const params = config.showParams ? paramsLine(photo.exif, "  /  ") : "";
  const date = config.showDate ? formatTakenAt(photo.exif.takenAt, config.dateFormat) : "";

  ctx.textAlign = "right";
  ctx.textBaseline = "alphabetic";
  if (params) {
    ctx.fillStyle = FUJI_INK;
    ctx.font = `700 ${Math.max(13, frameParams.fontSize * 1.2)}px ${fontFamily}`;
    ctx.fillText(params, rightX, captionTop + wordmarkSize * 0.85);
  }
  if (date) {
    ctx.fillStyle = FUJI_MUTED;
    ctx.font = `500 ${Math.max(11, frameParams.fontSize)}px ${fontFamily}`;
    ctx.fillText(date, rightX, captionTop + wordmarkSize * 0.85 + Math.max(20, frameParams.fontSize * 1.6));
  }
}

function drawPill(
  ctx: CanvasRenderingContext2D,
  x: number,
  y: number,
  w: number,
  h: number,
) {
  const r = h / 2;
  ctx.beginPath();
  ctx.moveTo(x + r, y);
  ctx.lineTo(x + w - r, y);
  ctx.arc(x + w - r, y + r, r, -Math.PI / 2, Math.PI / 2);
  ctx.lineTo(x + r, y + h);
  ctx.arc(x + r, y + r, r, Math.PI / 2, Math.PI * 1.5);
  ctx.closePath();
  ctx.fill();
}

File: src-tauri/src/render/templates/fujifilm_classic.rs
use image::{imageops::overlay, DynamicImage, Rgba, RgbaImage};

use crate::render::classic_bottom::{
    build_params_line, clean_display_text, compute_canvas_size, fill_rounded_rect_solid,
    fit_photo_in_area, format_camera, format_date, parse_color, resize_photo, ExportExif,
    ExportFrameParams, ExportTemplateConfig,
};
use crate::render::layout_spec::watermark_layout_spec;
use crate::render::text::TextRenderer;

const FUJI_GREEN: Rgba<u8> = Rgba([0, 100, 63, 255]);
const FUJI_PAPER: Rgba<u8> = Rgba([245, 241, 231, 255]);
const FUJI_INK: Rgba<u8> = Rgba([22, 22, 22, 255]);
const FUJI_MUTED: Rgba<u8> = Rgba([90, 86, 80, 255]);

pub(crate) fn compose_fujifilm_classic(
    source: DynamicImage,
    frame: &ExportFrameParams,
    exif: &ExportExif,
    config: &ExportTemplateConfig,
    text: Option<&TextRenderer>,
) -> RgbaImage {
    let src_w = source.width();
    let src_h = source.height();
    let (canvas_w, canvas_h) = compute_canvas_size(src_w, src_h, frame);
    let rs = src_w as f32 / 900.0;

    let side_margin = (canvas_w as f32 * 0.05).round() as u32;
    let top_margin = (canvas_h as f32 * 0.05).round() as u32;
    let caption_h = if config.show_watermark {
        ((canvas_h as f32 * 0.16).round() as u32).max((96.0 * rs) as u32)
    } else {
        top_margin
    };

    let area_w = canvas_w.saturating_sub(side_margin * 2);
    let area_h = canvas_h.saturating_sub(top_margin + caption_h);
    let (pw, ph, ox, oy) = fit_photo_in_area(area_w, area_h, src_w, src_h);
    let photo_x = (side_margin + ox) as i64;
    let photo_y = (top_margin + oy) as i64;
    let placed_x = side_margin + ox;
    let placed_y = top_margin + oy;

    let mut canvas = RgbaImage::from_pixel(canvas_w, canvas_h, FUJI_PAPER);

    let photo = resize_photo(source, pw, ph);
    overlay(&mut canvas, &photo, photo_x, photo_y);

    if !config.show_watermark {
        return canvas;
    }

    let Some(rend) = text else { return canvas; };

    let _ = watermark_layout_spec();
    // TS sizes: wordmarkSize = max(20, fontSize * 1.7); detail = max(11, fontSize); params = max(13, fontSize * 1.2)
    let wordmark_pt = (frame.font_size as f32 * 1.7).max(20.0) * rs;
    let detail_pt = (frame.font_size as f32).max(11.0) * rs;
    let params_pt = (frame.font_size as f32 * 1.2).max(13.0) * rs;
    let caption_gap = (canvas_h as f32 * 0.03).round();
    let caption_top = placed_y as f32 + ph as f32 + caption_gap;
    // TS uses alphabetic baseline; FUJIFILM baseline = caption_top + wordmark_pt * 0.85.
    let wordmark_baseline = caption_top + wordmark_pt * 0.85;
    let detail_gap = (frame.font_size as f32 * 1.6).max(20.0) * rs;
    let detail_baseline = wordmark_baseline + detail_gap;

    // Left: FUJIFILM wordmark
    rend.draw(
        &mut canvas,
        "FUJIFILM",
        placed_x as f32,
        wordmark_baseline - rend.ascent(wordmark_pt, true),
        wordmark_pt,
        true,
        FUJI_GREEN,
    );

    // Film simulation pill (first custom line or default)
    let sim_text = {
        let custom = config.custom_lines.first().cloned().unwrap_or_default();
        if custom.trim().is_empty() { "CLASSIC CHROME".to_string() } else { custom.trim().to_uppercase() }
    };
    let pill_pt = ((frame.font_size as f32) - 1.0).max(10.0) * rs;
    let sim_w = rend.measure(&sim_text, pill_pt, true);
    let pill_w = sim_w + pill_pt * 1.4;
    let pill_h = pill_pt * 1.9;
    let wordmark_w = rend.measure("FUJIFILM", wordmark_pt, true);
    let pill_x = placed_x as f32 + wordmark_w + pill_pt * 0.9;
    let pill_y = wordmark_baseline - pill_h * 0.78;
    fill_rounded_rect_solid(
        &mut canvas,
        pill_x as i64, pill_y as i64,
        pill_w as u32, pill_h as u32,
        (pill_h / 2.0) as u32,
        FUJI_GREEN,
    );
    let pill_text_x = pill_x + pill_w / 2.0 - sim_w / 2.0;
    let pill_text_y = pill_y + pill_h / 2.0 - rend.line_height(pill_pt, true) / 2.0;
    rend.draw(&mut canvas, &sim_text, pill_text_x, pill_text_y, pill_pt, true, Rgba([255, 255, 255, 255]));

    // Camera + lens below wordmark
    let camera = if config.show_camera { format_camera(exif) } else { String::new() };
    let lens = if config.show_lens { clean_display_text(&exif.lens) } else { String::new() };
    let detail: String = [camera, lens].into_iter().filter(|s| !s.is_empty()).collect::<Vec<_>>().join("  \u{00B7}  ");
    if !detail.is_empty() {
        let y = detail_baseline - rend.ascent(detail_pt, false);
        rend.draw(&mut canvas, &detail, placed_x as f32, y, detail_pt, false, FUJI_MUTED);
    }

    // Right: params + date
    let right_x = placed_x + pw;
    let params = if config.show_params { build_params_line(exif, "  /  ") } else { String::new() };
    let date = if config.show_date { format_date(&exif.taken_at, &config.date_format) } else { String::new() };

    if !params.is_empty() {
        let pw_meas = rend.measure(&params, params_pt, true);
        let y = wordmark_baseline - rend.ascent(params_pt, true);
        rend.draw(&mut canvas, &params, right_x as f32 - pw_meas, y, params_pt, true, FUJI_INK);
    }
    if !date.is_empty() {
        let dw = rend.measure(&date, detail_pt, false);
        let y = detail_baseline - rend.ascent(detail_pt, false);
        rend.draw(&mut canvas, &date, right_x as f32 - dw, y, detail_pt, false, FUJI_MUTED);
    }

    let _ = parse_color;
    canvas
}
